### DISCUSSION TITLE
Is there a way to apply accessibility identifier to an alert action button ?

### DISCUSSION THREAD

**Discussion by kristiyandobrev** [pointfreeco/swift-composable-architecture#2954] [Q&A]
Currently, our reducers utilize `AlertState` to manage alerts within our application. However, we've encountered a requirement to attach accessibility identifiers to the TextState within button actions. While we've noted the possibility of attaching an `accessibilityLabel`, we haven't found any mention of an `accessibilityIdentifier`.

I'm curious if there's a method I might have overlooked or if there's an actual approach for achieving this. Any insights would be greatly appreciated! 🤔

```swift
public static let customAlert = AlertState {
        TextState("Error")
    } actions: {
        ButtonState(role: .cancel, action: .closeButtonTapped) {
            TextState("Close").accessibilityLabel("closeButton")
        }
    } message: {
        TextState("Something went wrong")
}

**Comment by mbrandonw** [comment]
Hi @kristiyandobrev, can you confirm that even vanilla SwiftUI supports accessibility identifiers in buttons of alerts? There is a good chance that it does not.

The complicated thing with `accessibilityIdentifier` is that there isn't an overload on `Text` that returns `Text` as there is for `accessibilityLabel`:

```swift
let _: Text = Text("").accessibilityIdentifier("")  // ❌
let _: Text = Text("").accessibilityLabel("")       // ✅
```

So if vanilla SwiftUI _does_ support accessibility identifiers in alert buttons, then in order for us to support it we would need to turn to some type erasure. It probably wouldn't be a huge deal though.

But also at the end of the day you can always forgo `AlertState` and use the vanilla `.alert` view modifiers to construct the alert however you need. There's always going to be a limitation of just how much we can recreate with `AlertState` / `ButtonState` / `TextState`. It's just our best approximation of making those types testable, but we will probably never have it be a 100% faithful representation of SwiftUI's APIs.

**Comment by kristiyandobrev** [comment]
Hi @mbrandonw 

Thanks for the response ! 

It does seem like vanilla SwiftUI supports accessibility identifiers in buttons of alerts. 

As suggested we forwent the `AlertState` and took the vanilla SwiftUI path so that we are able to attach accessibility identifiers: 

```swift
 .alert(
    Text("Title"),
    isPresented: $store.showCancelAlert
) {
    Button(role: .none) {
        send(.confirmCancelButtonTapped)
    } label: {
        Text("Yes")
    }
    .accessibilityIdentifier("id1")

    Button(role: .cancel) {} label: {
        Text("No")
    }
    .accessibilityIdentifier("id2")
} message: {
    Text("Subtitle")
}
```

We can also verify the identifiers with the Accessibility Inspector tool.

Definitely preferring the `AlertState` and would be nice if you consider accessibility identifiers in future releases.
